### DISCUSSION TITLE
Usage guide

### DISCUSSION THREAD

**Discussion by zyloid** [aws/credentials-fetcher#38] [Q&A]
Hi, I'm trying to get Kerberos authentication working on Linux containers in AKS (Azure Kubernetes Service) and this looks like it could be very useful! I've got gMSA working on Windows containers and am keen to understand how this can be used to achieve the same thing on Linux. I have a few questions on usage if anyone is able to help out.

- What's the expected way that this should be deployed and used? The docs mention credentials-fetcher being an analogue for ccg.exe on Windows, however the documentation seems to imply that Linux hosts need to be domain joined - with ccg.exe the nodes do not need to be domain joined (which is a big benefit when cluster nodes are regularly scaled up and down). Would credentials-helper be expected to be deployed onto a domain-joined server outside the cluster that pods call into, or is the expectation that all cluster nodes are domain joined and run the credentials-helper daemon?

- My understanding is that the gMSA password resets every 30 days; does the credentials-fetcher daemon automatically regenerate the keytab files when this happens, or should the client application periodically call AddKerberosLease?

- With Windows pods gMSA is effectively the only option for Kerberos authentication. On Linux we have the option of generating and refreshing keytabs in a sidecar container. What's the benefit of using credentials-helper instead of a pregenerated keytab - is it simply that it allows us to use gMSA accounts instead of service accounts?

Any guidance would be much appreciated!

**Comment by smhmhmd** [comment]
Hi @zyloid 

Thanks for the starting the discussion.
Credentials-fetcher can be used in [domain-joined ](https://learn.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/manage-serviceaccounts#use-case-for-creating-gmsa-account-for-domain-joined-container-hosts)or [non-domain-joined mode](https://learn.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/manage-serviceaccounts#use-case-for-creating-gmsa-account-for-non-domain-joined-container-hosts). The non-domain-joined mode can be used with command line option as per this [commit](https://github.com/aws/credentials-fetcher/commit/67587826de447386ec55811b905d915b7d08f16b)

We can discuss your questions in detail, we would love to meet you.
Please email us: samiull at amazon . com , saakla at amazon . com , saikatba at amazon . com, we are in the  [PST time zone](https://greenwichmeantime.com/time-zone/usa/pacific-time/).

**Comment by smhmhmd** [comment]
Hi @zyloid ,

Added [issue](https://github.com/aws/credentials-fetcher/issues/39#issue-1506729329) for keytab support.
Please +1 the above issue.

**Comment by cschneiderwex** [comment]
Adding on here since I have a similar usage scenario as @zyloid. I am using aks and need to allow windows auth from linux containers to access sql server instances hosted on azure vms. This project seems like the way to do that. I am fairly new to kubernetes in general, so I am really unsure how to go about getting this working. Any step-by-step guidance would be appreciated. For some insight into my 'platform', the aks nodes are running ubuntu, our application containers are running .net core on debian I believe but we can change the base image to a different distro if needed. I'm really not sure if I need to create a daemonset in aks to run this and if so how do my pods/containers use it. Also if I need to build the source for a different linux distro like ubuntu, I'm not sure the steps to do that either. Any help would be greatly appreciated.

**Reply by smhmhmd** [reply]
Hi @cschneiderwex 

Just saw this.
Your use-case may not require keytab files, I think your use-case aligns with this PR - https://github.com/aws/credentials-fetcher/pull/53
For the PR, we tested a dotnet app (running in a EKS pod) that accesses SQL server using kerberos tickets (windows auth) created by credentials-fetcher.

**Reply by cschneiderwex** [reply]
Hello @smhmhmd,
Thanks for the reply. I think your link will be very helpful. I do have a couple of questions. 

1. In  step 6 about creating the secret.yaml file, where does the password value come from?
```shell
%cat secret1.yaml
apiVersion: v1
data:
  password: YWJjCg==
kind: Secret
metadata:
  name: krb-ticket1
``` 

2. I saw the issue someone raised about building credentials_fetcher for Ubuntu. Could you post or send me the solution that worked?

**Reply by smhmhmd** [reply]
Hi @cschneiderwex 

>In step 6 about creating the secret.yaml file, where does the password value come from?

The password value is just initialization, it may or may not be valid. However, when credentials-fetcher updates the password value (kerberos ticket), the value will be valid. We are using [this](https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/) to distribute kerberos tickets to pods from an instance running credentials-fetcher.
Credentials-fetcher does 'kube apply' periodically to refresh the secret.

>I saw the issue someone raised about building credentials_fetcher for Ubuntu. Could you post or send me the solution that worked?

We have a couple of remaining steps to make it work on ubuntu-20.04, the PR is in progress [here](https://github.com/aws/credentials-fetcher/compare/mainline...fordth:credentials-fetcher:mainline).

**Reply by cschneiderwex** [reply]
Great, thanks for the info.

**Reply by zhangxd6** [reply]
Do you have any updated step by step instruction to get it work in the k8s cluster?
